### PR TITLE
add support for DPT-Module V1

### PR DESCRIPTION
**Pull request by daanpape**
The DPT-Module V1 is an AR9331 SoM created by DPTechnics bvba. The
module comes with pre-flashed BlueCherry IoT keys on the flash hence
the custom partition layout. Detailed information about the module
can be found on the manufacturers website. Short specsheet:
- Atheros AR9331 400MHz MIPS24Kc
- 64MiB DDR2 RAM
- 16MiB Flash memory
- 150Mbps WiFi b/g/n 1T1R
- USB 2.0 Host
- 20 GPIO pins
- 2x 100Mbps ethernet (WAN/LAN)
- UART (Rx, Tx)
- SPI

Installing can be done through the on-board sysupgrade command:
1. Log in through telnet (no password by default)
2. Download the lede-ar71xx-generic-dpt-module-v1-squashfs-sysupgrade.bin
   file to the '/tmp' directory.
3. Use sysupgrade -n -v /tmp/lede-ar71xx-generic-dpt-module-v1-squashfs-
   sysupgrade.bin to start the upgrade to Lede.

Signed-off-by: Daan Pape <daan@dptechnics.com>

### CODE REVIEW

File: target/linux/ar71xx/base-files/etc/board.d/02_network [hunk 1]
```diff
@@ -29,6 +29,7 @@ ar71xx_setup_interfaces()
 	minibox-v1|\
 	mynet-n600|\
 	oolite|\
+	dpt-module-v1|\
```
Comment (blogic) [inline target/linux/ar71xx/base-files/etc/board.d/02_network]: alphabetical ordering please

File: target/linux/ar71xx/base-files/lib/ar71xx.sh [hunk 2]
```diff
@@ -293,6 +293,9 @@ tplink_board_detect() {
 	"3C0002"*)
 		model="MINIBOX_V1"
 		;;
+	"D00001"*)
+		model="DPT_MODULE_V1"
+		;;
 	"430000"*)
```
Comment (blogic) [inline target/linux/ar71xx/base-files/lib/ar71xx.sh]: numerical ordering

File: target/linux/ar71xx/base-files/lib/ar71xx.sh [hunk 3]
```diff
@@ -829,6 +832,9 @@ ar71xx_board_detect() {
 	*"Oolite V1.0")
 		name="oolite"
 		;;
+	*"DPT-Module V1")
+		name="dpt-module-v1"
+		;;
 	*"PB42")
```
Comment (blogic) [inline target/linux/ar71xx/base-files/lib/ar71xx.sh]: alphabetical ordering

File: target/linux/ar71xx/base-files/lib/upgrade/platform.sh [hunk 4]
```diff
@@ -393,6 +393,7 @@ platform_check_image() {
 	omy-x1|\
 	onion-omega|\
 	oolite|\
+	dpt-module-v1|\
 	re450|\
```
Comment (blogic) [inline target/linux/ar71xx/base-files/lib/upgrade/platform.sh]: alphabetical ordering

File: target/linux/ar71xx/config-4.4 [hunk 5]
```diff
@@ -121,6 +121,7 @@ CONFIG_ATH79_MACH_GL_MIFI=y
 CONFIG_ATH79_MACH_GL_USB150=y
 CONFIG_ATH79_MACH_GS_MINIBOX_V1=y
 CONFIG_ATH79_MACH_GS_OOLITE=y
+CONFIG_ATH79_MACH_DPT_MODULE_V1=y
```
Comment (blogic) [inline target/linux/ar71xx/config-4.4]: this symbol also needs to be added to the mikrotik and nand config as "is not set"

File: target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt [hunk 6]
```diff
@@ -847,6 +847,16 @@ config ATH79_MACH_GS_OOLITE
 	select ATH79_DEV_USB
 	select ATH79_DEV_WMAC
 
+config ATH79_MACH_DPT_MODULE_V1
```
Comment (blogic) [inline target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt]: alphabetical ordering

File: target/linux/ar71xx/files/arch/mips/ath79/Makefile [hunk 7]
```diff
@@ -127,6 +127,7 @@ obj-$(CONFIG_ATH79_MACH_GL_MIFI)		+= mach-gl-mifi.o
 obj-$(CONFIG_ATH79_MACH_GL_USB150)		+= mach-gl-usb150.o
 obj-$(CONFIG_ATH79_MACH_GS_MINIBOX_V1)		+= mach-gs-minibox-v1.o
 obj-$(CONFIG_ATH79_MACH_GS_OOLITE)		+= mach-gs-oolite.o
+obj-$(CONFIG_ATH79_MACH_DPT_MODULE_V1)		+= mach-dpt-module-v1.o
```
Comment (blogic) [inline target/linux/ar71xx/files/arch/mips/ath79/Makefile]: alphabetical ordering

File: target/linux/ar71xx/files/arch/mips/ath79/mach-dpt-module-v1.c [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+#include "dev-leds-gpio.h"
+#include "dev-m25p80.h"
+#include "dev-wmac.h"
+#include "machtypes.h"
+#include "dev-usb.h"
+
+#define DPT_MODULE_GPIO_BTN_RESET	       11                                   /* GPIO pin used for reset */
+#define DPT_MODULE_GPIO_LED_SYSTEM	       27                                   /* Green system LED */
+#define DPT_MODULE_KEYS_POLL_INTERVAL	   20                                   /* 20 milliseconds poll interval */
+#define DPT_MODULE_KEYS_DEBOUNCE_INTERVAL  (3 * DPT_MODULE_KEYS_POLL_INTERVAL)  /* Press is detected after 3 times same state */
```
Comment (blogic) [inline target/linux/ar71xx/files/arch/mips/ath79/mach-dpt-module-v1.c]: these 4 defines use tabs and not space between define and value

File: target/linux/ar71xx/files/arch/mips/ath79/machtypes.h [hunk 9]
```diff
@@ -119,6 +119,7 @@ enum ath79_mach_type {
 	ATH79_MACH_GL_USB150,			/* GL.iNet GL-USB150 */
 	ATH79_MACH_GS_MINIBOX_V1,		/* Gainstrong MiniBox V1.0 */
 	ATH79_MACH_GS_OOLITE,			/* GS OOLITE V1.0 */
+	ATH79_MACH_DPT_MODULE_V1,		/* DPTechnics DPT-Module V1 */
```
Comment (blogic) [inline target/linux/ar71xx/files/arch/mips/ath79/machtypes.h]: alphabetical ordering

File: target/linux/ar71xx/files/drivers/mtd/tplinkpart.c [hunk 10]
```diff
@@ -17,14 +17,17 @@
...
 #include <linux/version.h>
 
-#define TPLINK_NUM_PARTS	5
-#define TPLINK_HEADER_V1	0x01000000
-#define TPLINK_HEADER_V2	0x02000000
-#define MD5SUM_LEN		16
+#define TPLINK_NUM_PARTS        5
+#define TPLINK_HEADER_V1        0x01000000
+#define TPLINK_HEADER_V2        0x02000000
+#define MD5SUM_LEN              16
```
Comment (blogic) [inline target/linux/ar71xx/files/drivers/mtd/tplinkpart.c]: this removes tabs and adds spaces

File: target/linux/ar71xx/files/drivers/mtd/tplinkpart.c [hunk 11]
```diff
@@ -17,14 +17,17 @@
...
 
-#define TPLINK_ART_LEN		0x10000
-#define TPLINK_KERNEL_OFFS	0x20000
-#define TPLINK_64K_KERNEL_OFFS	0x10000
+#define TPLINK_ART_LEN          0x10000
+#define TPLINK_KERNEL_OFFS      0x20000
+#define TPLINK_64K_KERNEL_OFFS  0x10000
```
Comment (blogic) [inline target/linux/ar71xx/files/drivers/mtd/tplinkpart.c]: this removes tabs and adds spaces

File: target/linux/ar71xx/image/generic.mk [hunk 12]
```diff
@@ -721,6 +721,17 @@ define Device/tellstick-znet-lite
 endef
 TARGET_DEVICES += tellstick-znet-lite
 
+define Device/dpt-module-v1
```
Comment (blogic) [inline target/linux/ar71xx/image/generic.mk]: alphabetical ordering

File: target/linux/ar71xx/mikrotik/config-default [hunk 13]
```diff
@@ -85,6 +85,7 @@
 # CONFIG_ATH79_MACH_GL_USB150 is not set
 # CONFIG_ATH79_MACH_GS_MINIBOX_V1 is not set
 # CONFIG_ATH79_MACH_GS_OOLITE is not set
+# CONFIG_ATH79_MACH_DPT_MODULE_V1 is not set
```
Comment (blogic) [inline target/linux/ar71xx/mikrotik/config-default]: alphabetical ordering

File: target/linux/ar71xx/nand/config-default [hunk 14]
```diff
@@ -86,6 +86,7 @@ CONFIG_ATH79_MACH_DOMYWIFI_DW33D=y
 # CONFIG_ATH79_MACH_GL_USB150 is not set
 # CONFIG_ATH79_MACH_GS_MINIBOX_V1 is not set
 # CONFIG_ATH79_MACH_GS_OOLITE is not set
+# CONFIG_ATH79_MACH_DPT_MODULE_V1 is not set
```
Comment (blogic) [inline target/linux/ar71xx/nand/config-default]: alphabetical ordering

**Comment by daanpape** [comment]
@blogic Thank you for your reviews. I'm quite new to making pull requests. Should I close this one and make a new one or can I fix the remarks in this pull request? Kind regards, Daan

**Comment by blogic** [comment]
@daanpape no worries everyone needs to start somewhere. the best way is to just make the changes then "git commit --amend" them to your current tree and then do a forced push "git push -f" to push the new amended patch to this PR.

**Comment by blogic** [comment]
With the remerge in progress, all PRs on the lede-project organisation will be closed. Please help getting this merged or rebase/post it on the openwrt project page (https://github.com/openwrt/openwrt/pulls). All remaining PRs will be closed in 30 days.

**Comment by pepe2k** [comment]
Hi @daanpape,

> The DPT-Module V1 is an AR9331 SoM created by DPTechnics bvba. [...]

This doesn't seem to be true. Your module looks just like a rebranded [Gainstrong Oolite v 1](https://wiki.openwrt.org/toh/oolite/oolitev1) (FCC ID: [2AB5EGS-OOLITE](https://fccid.io/2AB5EGS-OOLITE)) which is already supported in OpenWrt. For reference: [DPT-Module V1](https://dptechnics.com/en/products/dpt-module-v1.html) vs.  [Gainstrong Oolite v 1](http://oolite.cn/category/qualcomm/ar9331/oolite-v1-0).

You can drop all that redundant mach-*.c code and use what is already there for Oolite. And as far as I can see, you can also drop that weird mtd parser and just use static mtd definition.

Cheers,
Piotr

**Comment by daanpape** [comment]
@pepe2k Maybe I wasn't clear with the DPT-Module description. This is actually used for DPT-Module based hardware such as the DPT-Board. This hardware is very different of the OOLITE hardware, also the module is not a 1-1 copy but uses a different flash layout. 

How can I add the different flash layout without the mach-*.c code? How is the static mtd definition done? I'm working on getting  this code merged before the 30-day limit passes.

**Comment by pepe2k** [comment]
Hi @daanpape,

> Maybe I wasn't clear with the DPT-Module description. This is actually used for DPT-Module based hardware such as the DPT-Board.

Then maybe you should add support for DPT-Board, not just for the module.

> This hardware is very different of the OOLITE hardware, also the module is not a 1-1 copy but uses a different flash layout.

Can you tell us FCC ID of your module?

> How can I add the different flash layout without the mach-*.c code? How is the static mtd definition done? I'm working on getting this code merged before the 30-day limit passes.

Have a look how it's done here: https://github.com/lede-project/source/blob/master/target/linux/ar71xx/image/generic.mk. You would still need some chunks of code in mach-*.c, like different LED names, but you can include your changes in `mach-gs-oolite.c`.

Cheers,
Piotr

**Pull request closed without merging**